### PR TITLE
Feature: OpenID Connect Discovery support

### LINKED ISSUE #241: Open ID Connect Discovery support
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

I'm thinking about supporting [OpenID connect discovery](https://openid.net/specs/openid-connect-discovery-1_0.html) for this plugin.
`simple-oauth2` requires static host/path combinations to be configured as options, for someone that is using this library it might be annoying to configure everything. 

So idea is something like I've shown in [this comment](https://github.com/fastify/fastify-oauth2/pull/206#discussion_r1399712774). 
Essentially, send GET to discovery endpoint and then bake options based on the response that will fit into `credentials.auth` for the plugin.

Let me know what do you think how this API could look like based on example I'm giving below.

The only eventual drawback I see is that user needs to understand that GET request happens there, 
and that they have to now nest their `fastify.listen` call into continuation fn.

### Motivation

This plugin does best effort in providing some well known providers as well as options to support custom provider, but it still requires configuration of like 5 params. 
Often developers get it wrong, and it leads to invalid configs, false bug reports etc...

### Example

```js
const oauth2Plugin = require('fastify-oauth2')
const issuer = process.env.ISSUER

oauth2Plugin.discover(issuer)
  .then(metadata => {
    const pluginOpts = {
      credentials:  {
        auth: metadata,
        // ...other credential opts...
        client_id: ...
        client_secret: ...
      }
      // other opts
    }
    fastify.register(oauthPlugin, pluginOpts)
    fastify.listen({ port: 3000 })
  })
  .catch(err => {
    // handle discovery err
    
  })
```

### PR DESCRIPTION
**Pull request by big-kahuna-burger**
Closes #241

New option `discovery` is now available as described in #241.
Usage is very simple:
- add `discovery` object with `issuer` url in it
- remove `credentials.auth` to avoid confusion related to what is used, plugin will throw error down if both are used (since the intent is unclear)
- it will also select appropriate PKCE method based on metadata response when possible, 
(but if you want to force explicit method, you can still do it using top level option `pkce`)

Video shows how easy it is to switch from one issuer to another (I'm here switching from Google to Local OIDC provider, that has same client_id/client_secret pair :))

https://github.com/fastify/fastify-oauth2/assets/11753867/c36e464f-efea-49ef-b4fb-a90f63482270

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by big-kahuna-burger** [comment]
```sh
fastify-oauth2 git:(discovery) npm test

> @fastify/oauth2@7.5.0 test
> npm run test:unit && npm run test:typescript


> @fastify/oauth2@7.5.0 test:unit
> tap

​ PASS ​ test/index.test.js 233 OK 296.342ms


                         
  🌈 SUMMARY RESULTS 🌈  
                         

Suites:   ​1 passed​, ​1 of 1 completed
Asserts:  ​​​233 passed​, ​of 233
​Time:​   ​547.579ms
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |     100 |      100 |     100 |     100 |                   
 index.js |     100 |      100 |     100 |     100 |                   
----------|---------|----------|---------|---------|-------------------

> @fastify/oauth2@7.5.0 test:typescript
> tsd
```

**Comment by big-kahuna-burger** [comment]
@mcollina I'm not sure what's wrong with Github Actions here

cc @panva
Pinging Filip to give it a look for correctness as well.

**Comment by panva** [comment]
Apologies, I'm not going to be able to review this index diff.

**Comment by big-kahuna-burger** [comment]
> Apologies, I'm not going to be able to review this index diff.

Understandable. 
I had to nest the plugin config internally, that's why the diff is large.

File: index.js [hunk 1]
```diff
@@ -83,182 +86,301 @@ function fastifyOauth2 (fastify, options, next) {
...
   }
+}
 
-  try {
-    fastify.decorate(name, decoration)
-    fastify.decorate(`oauth2${name.slice(0, 1).toUpperCase()}${name.slice(1)}`, decoration)
-  } catch (e) {
-    next(e)
-    return
+function selectPkceFromMetadata (metadata) {
```
Comment (panva) [inline index.js:359]: Always use S256 PKCE even when code_challenge_methods_supported is not given.

Comment (big-kahuna-burger) [reply index.js:359]: AS should ignore it if doesn't support it, right?

**Comment by panva** [comment]
please remove me as pending reviewer, I don't have the bandwith for this these days

**Comment by big-kahuna-burger** [comment]
I think that's on @mcollina. 
Thanks for the part you did. Appreciate that.

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**